File: Aula AG 1/ag_main.py
from cromossomo import Cromossomo

if __name__ == '__main__':
    tamanho_populacao = int(input('Tamanho da população: '))
    taxa_selecao = int(input('Taxa de seleção [25 a 40%]: '))
    taxa_reproducao = 100 - taxa_selecao
    taxa_mutacao = int(input('Taxa de mutação [3 a 10%]: '))
    quantidade_geracoes = int(input('Quantidade de gerações: '))
                            
    populacao = []
    nova_populacao = []

    Cromossomo.gerar_populacao(populacao, tamanho_populacao)
    populacao.sort(key=lambda cromossomo: cromossomo.aptidao)
    print('Geração 0...')
    Cromossomo.exibir_populacao(populacao)

    for i in range(1, quantidade_geracoes + 1):
        Cromossomo.selecionar_por_torneio(populacao, nova_populacao, taxa_selecao)
        Cromossomo.reproduzir(populacao, nova_populacao, taxa_reproducao)
        if i % taxa_mutacao == 0:
            Cromossomo.mutar(nova_populacao)

        populacao = nova_populacao[:]
        nova_populacao.clear()
        populacao.sort(key=lambda cromossomo: cromossomo.aptidao)
        print(f'Geração {i}...')
        Cromossomo.exibir_populacao(populacao)

File: Atividade cidades java/AGMain.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

public class AGMain {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Tamanho da população: ");
        int tamanhoPopulacao = scanner.nextInt();
        
        System.out.print("Taxa de seleção [25 a 40%]: ");
        int taxaSelecao = scanner.nextInt();
        int taxaReproducao = 100 - taxaSelecao;
        
        System.out.print("Taxa de mutação [3 a 10%]: ");
        int taxaMutacao = scanner.nextInt();
        
        System.out.print("Quantidade de gerações: ");
        int quantidadeGeracoes = scanner.nextInt();

        List<Cromossomo> populacao = new ArrayList<>();
        List<Cromossomo> novaPopulacao = new ArrayList<>();

        Cromossomo.gerarPopulacao(populacao, tamanhoPopulacao);
        populacao.sort(Comparator.comparingInt(Cromossomo::getAptidao));
        System.out.println("Geração 0...");
        Cromossomo.exibirPopulacao(populacao);

        for (int i = 1; i <= quantidadeGeracoes; i++) {
            Cromossomo.selecionarPorTorneio(populacao, novaPopulacao, taxaSelecao);
            Cromossomo.reproduzir(populacao, novaPopulacao, taxaReproducao);
            
            if (i % taxaMutacao == 0) {
                Cromossomo.mutar(novaPopulacao);
            }

            populacao = new ArrayList<>(novaPopulacao);
            novaPopulacao.clear();
            populacao.sort(Comparator.comparingInt(Cromossomo::getAptidao));
            
            System.out.println("Geração " + i + "...");
            Cromossomo.exibirPopulacao(populacao);
        }
        scanner.close();
    }
}